File: kotlin/fizzbuzz-outside-in/src/main/kotlin/FizzBuzz.kt
class FizzBuzz(private vararg val rules: Rule = arrayOf(
        ConcreteRule(isDivisibleBy(3), "Fizz"),
        ConcreteRule(isDivisibleBy(5), "Buzz")
)) {
    fun calculate(input: Int): String = rules
            .filter { it.appliesTo(input) }
            .joinToString("-") { it.result }
            .orWhenEmpty { input.toString() }

    private fun String.orWhenEmpty(alternative: () -> String): String = when {
        isEmpty() -> alternative()
        else -> this
    }
}

private fun isDivisibleBy(d: Int) = { input: Int -> input % d == 0 }

private class ConcreteRule(val applies: (Int) -> Boolean, override val result: String) : Rule {
    override fun appliesTo(input: Int): Boolean = applies(input)
}

File: typescript/express/src/services/fizz-buzz.ts
export class FizzBuzz {
    private rules = [
        {appliesTo: this.isDivisibleBy(3), result: 'Fizz'},
        {appliesTo: this.isDivisibleBy(5), result: 'Buzz'},
    ]

    calculate(input: number): string {
        return this.rules
            .filter(r => r.appliesTo(input))
            .map(r => r.result)
            .join('-') || input.toString()
    }

    private isDivisibleBy(divisor: number) {
        return (input: number) =>
            input % divisor === 0
    }
}
